governance/pyth_staking_sdk/src/pyth-staking-client.ts

```diff
@@ -1031,4 +1037,114 @@ export class PythStakingClient {
 
     return getAccount(this.connection, rewardCustodyAccountAddress);
   }
+
+  /**
+   * Return all stake account positions for all owners.  Note that this method
+   * is unique in a few ways:
+   *
+   * 1. It's very, very expensive.  Don't call it if you don't _really_ need it,
+   *    and expect it to take a few minutes to respond.
+   * 2. Because the full positionData is so large, json parsing it with a
+   *    typical json parser would involve buffering to a string that's too large
+   *    for node.  So instead we use `stream-json` to parse it as a stream.
+   */
+  public async getAllStakeAccountPositionsAllOwners(): Promise<
+    StakeAccountPositions[]
+  > {
+    const res = await fetch(this.connection.rpcEndpoint, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify({
+        jsonrpc: "2.0",
+        id: 1,
+        method: "getProgramAccounts",
+        params: [
+          this.stakingProgram.programId.toBase58(),
+          {
+            encoding: "base64",
+            filters: [
+              {
+                memcmp: this.stakingProgram.coder.accounts.memcmp(
+                  "positionData",
+                ) as {
+                  offset: number;
+                  bytes: string;
+                },
+              },
+            ],
+          },
+        ],
+      }),
+    });
+
+    if (res.ok) {
+      const { body } = res;
+      if (body) {
+        const accounts = await new Promise<unknown[]>((resolve) => {
+          const pipeline = chain([
+            body,
+            parser(),
+            pick({ filter: "result" }),
+            ignore({ filter: /(owner|executable|lamports|rentEpoch)/i }),
+            streamArray(),
+          ]);
+
+          const accounts: unknown[] = [];
+          pipeline.on("data", (data: unknown) => accounts.push(data));
+          pipeline.on("end", () => {
+            resolve(accounts);
+          });
+        });
+
+        return accountSchema
+          .parse(accounts)
+          .map(({ value }) =>
+            deserializeStakeAccountPositions(
+              value.pubkey,
+              value.account.data,
+              this.stakingProgram.idl,
+            ),
+          );
+      } else {
+        throw new NoBodyError();
+      }
+    } else {
+      throw new NotOKError(res);
+    }
+  }
+}
+
+const accountSchema = z.array(
+  z.object({
+    value: z.object({
+      account: z.object({
+        data: z
+          .array(z.string())
+          .min(1)
+          .transform((data) =>
+            // Safe because `min(1)` guarantees that `data` is nonempty
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            Buffer.from(data[0]!, "base64"),
```

Comment:
nit: can we use an if to avoid having !?

Reply:
In this case we actually want `!` because the `min(1)` line from Zod already guarantees the length, but can't encode that in the type.  I generally prefer letting Zod verify this kind of thing since it will generate nice consistent error messages, but it requires the `!` to avoid having a second null check at runtime.  In this case since it's so many accounts, it's not huge but it is actually a relevant optimization.

I usually wrap this kind of thing into a `nonEmptyArray` Zod helper which returns `[T, ...T[]]` instead of `T[]` which typescript can then guarantee is nonempty...